dcos_test_utils/gce.py

```diff
@@ -0,0 +1,247 @@
+
+
+def catch_http_exceptions(f):
+    '''
+    Runs the function "f" and raises an exception with a custom help message if a google HttpError with status 404 or
+        409 is caught
+    :param f: function to be executed
+    :return: return value of function "f"
+    '''
+    def handle_exception(*args, **kwargs):
+        try:
+            return f(*args, **kwargs)
+        except HttpError as e:
+            if e.resp.status == 404:
+                log.exception("The resource you are trying to access doesn't exist")
+            elif e.resp.status == 409:
+                log.exception('''The specified resources exist and might be under an active operation
+                                   (operation conflict)''')
+            raise e
+
+    return handle_exception
+
+
+class GceWrapper:
+    @catch_http_exceptions
+    def __init__(self, credentials_dict):
+        credentials = GoogleCredentials.get_application_default()
+        self.compute = discovery.build('compute', 'v1', credentials=credentials)
+        self.deployment_manager = discovery.build('deploymentmanager', 'v2', credentials=credentials)
+        self.project_id = credentials_dict['project_id']
+
+    def build_template_body(self, config) -> str:
+        node_count = 1 + config['num_masters'] + config['num_public_agents'] + config['num_private_agents']
+        template_body = 'resources:'
+        for i in range(node_count):
+            template_body += TEMPLATE.format(name='vm' + str(i), sourceImage=config['source_image'],
+                                             machineType=config['machine_type'], imageProject=config['image_project'],
+                                             project_id=self.project_id, zone=config['zone'], )
+        return template_body
+
+    @catch_http_exceptions
+    def deploy_instances(self, config) -> dict:
+        config['template_body'] = self.build_template_body(config)
+
+        body = {
+            'name': config['deployment_name'],
+            'target': {
+                'config': {
+                    'content': config['template_body']
+                }
+            }
+        }
+
+        return self.deployment_manager.deployments().insert(project=self.project_id, body=body).execute()
+
+    @catch_http_exceptions
+    def get_instances_info(self, deployment: Deployment) -> [dict]:
+        request = self.compute.instances().list(project=self.project_id, zone=deployment.zone)
+        while request is not None:
+            response = request.execute()
+            for instance_info in response['items']:
+                yield instance_info
+            request = self.deployment_manager.resources().list_next(previous_request=request,
+                                                                    previous_response=response)
+
+    @catch_http_exceptions
+    def get_deployment_info(self, deployment: Deployment) -> dict:
+        return self.deployment_manager.deployments().get(project=self.project_id,
```

Comment:
(project_id, deployment_name) -> globally unique pair
The methods that do require the zone get and set info about instances, and instances have the following globally unique combo: (project_id, instance_name, zone)
Which brings up something that needs to be fixed: this means there can be name instance_name clashes between deployments